network/default.go

```diff
@@ -439,11 +439,42 @@ func (n *network) processCtrlChan(l tunnel.Listener) {
 			case "advert":
 				pbRtrAdvert := &pbRtr.Advert{}
 				if err := proto.Unmarshal(m.Body, pbRtrAdvert); err != nil {
+					log.Debugf("Network fail to unmarshal advert message: %v", err)
 					continue
 				}
 
+				// loookup advertising node in our neighbourhood
+				n.RLock()
+				advertNode, ok := n.neighbours[pbRtrAdvert.Id]
+				if !ok {
+					// advertising node has not been registered as our neighbour, yet
+					// let's add it to the map of our neighbours
+					advertNode = &node{
+						id:         pbRtrAdvert.Id,
+						neighbours: make(map[string]*node),
+					}
+					n.neighbours[pbRtrAdvert.Id] = advertNode
+				}
+				n.RUnlock()
+
 				var events []*router.Event
 				for _, event := range pbRtrAdvert.Events {
+					// set the address of the advertising node
+					// we know Route.Gateway is the address of advertNode
+					// NOTE: this is true only when advertNode had not been registered
+					// as our neighbour when we received the advert from it
+					if advertNode.address == "" {
+						advertNode.address = event.Route.Gateway
+					}
+					// if advertising node id is not the same as Route.Router
+					// we know the advertising node is not the origin of the route
+					if advertNode.id != event.Route.Router {
+						// if the origin router is not in the advertising node neighbourhood
+						// we can't rule out potential routing loops so we bail here
+						if _, ok := advertNode.neighbours[event.Route.Router]; !ok {
+							continue
```

Comment:
A secondary check here might be necessary. 

If event.Route.Router is in your neighbour list, check if the advert node is in that list. If it's not then you can also ignore.

Reply:
Yeah, thats a good point (checking bidirectionality). I actually had that in my original code before I changed that back to this version and I can't remember now why >_<

Reply:
Oh I know why. I figured if the `event.Route.Router` is in `advertNode` neighbour list that means we can route to `Route.Router` directly from `advertNode`, hence no need to check the other direction. Felt redundant -- as sooner or later they'll announce themselves to each other.

Reply:
Wait, here's a question. Does that mean we limit routes that we know of to one hop from our neighbour? Meaning metric of 100. Should we be doing this programmatically by defining a hop count so we actually get to that metric of 1000? Or is this a secondary concern? We know we don't want all the routes but how many do we want?

Reply:
I think having a hop count is an interesting one. So, we can tell hop count up to 2 from the direct neighbourhood list and neighbours' neighbourhood. That is reflected in calculating the Metric in `setRouteMetric`: anything beyond hop count 2 is assigned Metric 1000

Think the idea of allowing a certain size of neighbourhood is an interesting thought to consider.

Reply:
I'll make a note rather than adding a card for this and merge the current PR.